libpromises/evalfunction.c

```diff
@@ -1331,79 +1349,159 @@ static FnCallResult FnCallPackagesMatching(ARG_UNUSED EvalContext *ctx, ARG_UNUS
-                "Line %d from package inventory '%s' is too long (%zd) to be sensible",
-                linenumber, filename, strlen(line));
             free(line);
-            break; /* or continue ? */
+            break;
         }
+        free(line);
+    }
 
-        if (StringMatchFullWithPrecompiledRegex(matcher, line))
+    bool ret = feof(fin);
+    fclose(fin);
+    
+    return ret;
+}
+
+static bool GetPackagesMatching(pcre *matcher, JsonElement *json, const bool installed_mode, Rlist *default_inventory)
+{
+    dbid database = (installed_mode == true ? dbid_packages_installed : dbid_packages_updates);
+
+    for (const Rlist *rp = default_inventory; rp != NULL; rp = rp->next)
+    {
+        const char *pm_name =  RlistScalarValue(rp);
+        size_t pm_name_size = strlen(pm_name);
+        
+        Log(LOG_LEVEL_DEBUG, "Reading packages (%d) for package modeule [%s]", 
+                database, pm_name);
+        
+        if (StringSafeEqual(pm_name, "cf_null"))
         {
-            Seq *list = SeqParseCsvString(line);
-            if (SeqLength(list) != 4)
+            continue;
+        }
+        
+        CF_DB *db_cached;
+        if (!OpenSubDB(&db_cached, database, pm_name))
+        {
+            Log(LOG_LEVEL_ERR, "Can not open database %d to get packages data.", database);
+            return false;
+        }
+
+        char *key = "<inventory>";
+        int data_size = ValueSizeDB(db_cached, key, strlen(key) + 1);
+        
+        Log(LOG_LEVEL_DEBUG, "Reading inventory from database: %d", data_size);
+        
+        /* For empty list we are storing one byte value in database. */
+        if (data_size > 1)
+        {
+            char *buff = xmalloc(data_size + 1);
+            buff[data_size] = '\0';
+            if (!ReadDB(db_cached, key, buff, data_size))
             {
-                Log(LOG_LEVEL_ERR,
-                    "Line %d from package inventory '%s' did not yield 4 elements: %s",
-                    linenumber, filename, line);
-                ++linenumber;
-                SeqDestroy(list);
-                free(line);
+                Log(LOG_LEVEL_WARNING, "Can not read installed packages database "
+                    "for '%s' package module.", pm_name);
                 continue;
             }
+            
+            Seq *packages_form_module = SeqStringFromString(buff, '\n');
+            free(buff);
+            
+            if (packages_form_module)
+            {
+                // Iterate over and see where match is.
+                for (int i = 0; i < SeqLength(packages_form_module); i++)
+                {
+                    char *line = SeqAt(packages_form_module, i);
+                    size_t new_line_size = strlen(line) + pm_name_size + 2; // we need coma and terminator
+                    char new_line[new_line_size];
+                    strcpy(new_line, line);
+                    strcat(new_line, ",");
+                    strcat(new_line, pm_name);
```

Comment:
I will add a comment, but yes. The packages are spitted by the comma. 
In the old package promise we've been using CSVs for reporting, and I left it to make sure that it will be easy to integrate this into all places when we are using package promise. This is one place, where this decisions is paying off.